aezeed/cipherseed.go

```diff
@@ -0,0 +1,542 @@
+	// aez.
+	keyLen = 32
+
+	// bitsPerWord is the number of bits each word in the wordlist encodes.
+	// We encode our mnemonic using 24 words, so 264 bits (33 bytes).
+	bitsPerWord = 11
+
+	// saltOffset is the index within an enciphered cipherseed that marks
+	// the start of the salt.
+	saltOffset = EncipheredCipherSeedSize - checkSumSize - saltSize
+
+	// checkSumSize is the index within an enciphered cipher seed that
+	// marks the start of the checksum.
+	checkSumOffset = EncipheredCipherSeedSize - checkSumSize
+
+	// encipheredSeedSize is the size of the cipherseed before applying the
+	// external version, salt, and checksum for the final encoding.
+	encipheredSeedSize = DecipheredCipherSeedSize + CipherTextExpansion
+)
+
+var (
+	// Below at the default scrypt parameters that are tied to
+	// CipherSeedVersion zero.
+	scryptN = 32768
+	scryptR = 8
+	scryptP = 1
+
+	// crcTable is a table that presents the polynomial we'll use for
+	// computing our checksum.
+	crcTable = crc32.MakeTable(crc32.Castagnoli)
+)
+
+var (
+	// bitcoinGenesisDate is the timestamp of Bitcoin's genesis block.
+	// We'll use this value in order to create a compact birthday for the
+	// seed. The birthday will be interested as the number of days since
+	// the genesis date. We refer to this time period as ABE (after Bitcoin
+	// era).
+	bitcoinGenesisDate = time.Unix(1231006505, 0)
+)
+
+// CipherSeed is a fully decoded instance of the aezeed scheme. At a high
+// level, the encoded cipherseed is the enciphering of: a version byte, a set
+// of bytes for a timestamp, the entropy which will be used to directly
+// construct the HD seed, and finally a checksum over the rest. This scheme was
+// created as the widely used schemes in the space lack two critical traits: a
+// version byte, and a birthday timestamp. The version allows us to modify the
+// details of the scheme in the future, and the birthday gives wallets a limit
+// of how far back in the chain they'll need to start scanning. We also add an
+// external version to the enciphering plaintext seed. With this addition,
+// seeds are able to be "upgraded" (to diff params, or entirely diff crypt),
+// while maintaining the semantics of the plaintext seed.
+//
+// The core of the scheme is the usage of aez to carefully control the size of
+// the final encrypted seed. With the current parameters, this scheme can be
+// encoded using a 24 word mnemonic. We use 4 bytes of ciphertext expansion
+// when enciphering the raw seed, giving us the equivalent of 40-bit MAC (as we
+// check for a particular seed version). Using the external 4 byte checksum,
+// we're able to ensure that the user input the correct set of words.  Finally,
+// the password in the scheme is optional. If not specified, an empty string
+// will be used as the password. Otherwise, the addition of the password means
+// that users can encrypt the saw "plaintext" seed under distinct passwords to
```

Comment:
I don't get this... That the users can encrypt the plaintext unders distinct passwords, is that a desirable feature?

Reply:
Yes, this means they can replicate the _same_ seed widely under distinct passwords. Each one when decrypted will allow then to restore their wallet. 

One can even take an existing seed, and then produce a new one with a diff passphrase.